Set model to orbit-fit when setting to StageMode

Phipson opened the pull request:
#### 256e7b32f25c8ce4c6180576dfdb3f6c888b847a
<pre>
Set model to orbit-fit when setting to StageMode
<a href="https://bugs.webkit.org/show_bug.cgi?id=290136">https://bugs.webkit.org/show_bug.cgi?id=290136</a>
<a href="https://rdar.apple.com/146901729">rdar://146901729</a>

Reviewed by Ada Chan.

This PR adjusts the way we calculate the bounding box and center of the model based
on whether it is in StageMode orbit or not. When in orbit, we want to inset the model
based on the bounding sphere instead of the bounding box, such that as we rotate the
model it does not get clipped. Since the model&apos;s entityTransform is read-write when
we transition back to StageMode none, we do not re-compute when we exit StageMode
orbit. We let web developers recompute the transform as they please.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::computeSRT):
- Added a boolean to set either a default rotation (if we just loaded the model or
moved it outside the portal); or reuse the existing rotation of the model
- Also added a separate compute path to use the boundingRadius of the model
(WebKit::ModelProcessModelPlayerProxy::computeTransform):
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::setHasPortal):
(WebKit::ModelProcessModelPlayerProxy::setStageMode):
- Updated callbacks for computeTransform
* Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
(WKSRKEntity.boundingRadius):
- Returns bounding radius of the model
(WKSRKEntity.recenterEntity(at:)):
- Recenters the model such that the interaction pivot point is at the center
of the model for orbit fit
* Source/WebKit/WebKitSwift/RealityKit/RealityKitBridging.h:

Canonical link: <a href="https://commits.webkit.org/292749@main">https://commits.webkit.org/292749@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/517d38a13aae42a31460ee885c30ef62efc3b3aa

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97019 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16642 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6842 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102100 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47544 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16934 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25092 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73905 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31119 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100022 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12769 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87747 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54241 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12521 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5596 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46874 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82597 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5675 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104121 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24093 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/17576 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82954 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24344 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83848 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82348 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20702 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27036 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4578 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/17592 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24058 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29206 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23881 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27193 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25454 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -321,31 +321,39 @@ static void computeScaledExtentsAndCenter(simd_float2 boundsOfLayerInMeters, sim
+            srt.translation = simd_make_float3(-boundingBoxCenter.x, -boundingBoxCenter.y, -boundingBoxCenter.z - boundingSphereDiameter * minScale / 2.0f);
+        else
+            srt.translation = simd_make_float3(-boundingBoxCenter.x, -boundingBoxCenter.y, -boundingBoxCenter.z + boundingSphereDiameter * minScale / 2.0f);
+    }
```
It's something we should have been doing before - but let's add some checks to the math (both none and orbit cases) to make sure we won't divide by zero.

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -704,16 +720,7 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
-        WebCore::TransformationMatrix transform = WebCore::TransformationMatrix(matrix);
-        send(Messages::ModelProcessModelPlayer::DidUpdateEntityTransform(transform));
+        updateTransformSRT();
     }
```
Nit: I think our coding style is to not use curly braces around single line if blocks.

achan00 reviewed (approved):
Thanks for adding the test!

Phipson committed:
Set model to orbit-fit when setting to StageMode
https://bugs.webkit.org/show_bug.cgi?id=290136
rdar://146901729

Reviewed by Ada Chan.

This PR adjusts the way we calculate the bounding box and center of the model based
on whether it is in StageMode orbit or not. When in orbit, we want to inset the model
based on the bounding sphere instead of the bounding box, such that as we rotate the
model it does not get clipped. Since the model's entityTransform is read-write when
we transition back to StageMode none, we do not re-compute when we exit StageMode
orbit. We let web developers recompute the transform as they please.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::computeSRT):
- Added a boolean to set either a default rotation (if we just loaded the model or
moved it outside the portal); or reuse the existing rotation of the model
- Also added a separate compute path to use the boundingRadius of the model
(WebKit::ModelProcessModelPlayerProxy::computeTransform):
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::setHasPortal):
(WebKit::ModelProcessModelPlayerProxy::setStageMode):
- Updated callbacks for computeTransform
* Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
(WKSRKEntity.boundingRadius):
- Returns bounding radius of the model
(WKSRKEntity.recenterEntity(at:)):
- Recenters the model such that the interaction pivot point is at the center
of the model for orbit fit
* Source/WebKit/WebKitSwift/RealityKit/RealityKitBridging.h:

Canonical link: https://commits.webkit.org/292749@main
